### PR TITLE
Add a section about binary size reduction

### PR DESCRIPTION
**Pull request by j8r**
~~This adds instructions for static linking.
I've also included a little change to the link explaining heap/stack memory.~~
- Improve the title to reflect the dynamically-linked nature of the compiled executable 
- Create a section about binary size reduction

### CODE REVIEW

**Comment by straight-shoota** [comment]
I'm not sure that experimental behaviour (even if labeled as such) should be in the language reference. Of course, Crystal itself is still somewhat experimental and many things might get changed before 1.0 is released. But there is still a difference to highly experimental stuff.

Maybe this could better be detailed in the Github Wiki for the time being.

The unrelated changes are good. The updated compiler output is already in #178 . The SO link should probably be in a separate PR.

**Comment by j8r** [comment]
Ok I will rework this PR. I agree, we can keep the static linking section, explaining the `--static` flag, and add a link to a FAQ in the GitHub wiki to answer differents issues that can be faced.

File: using_the_compiler/README.md [hunk 1]
```diff
@@ -44,6 +44,25 @@ Make sure to always use `--release` for production-ready executables and when pe
...
 
+To reduce the binary size, you can add the `--no-debug` flag and use the `strip` command. Debug symbols will be removed, use this options if only the size is an issue and you won't need to debug the program.
+
+## Creating a statically-linked executable
+
+This following libraries need to be required in the programm to force the linking against `libstdc++`:
+
+```
+require "llvm/lib_llvm"
+require "llvm/enums"
```
Comment (RX14) [inline using_the_compiler/README.md]: Very heavily -1 on documenting a bug.

Comment (j8r) [reply using_the_compiler/README.md]: I guess we will add this to the wiki, I remove it.

**Comment by j8r** [comment]
I've removed the dirty hacks related to the static compilation @RX14 . I have kept them in another place temporaly on my side. Do we have to create a new Wiki page, or adding this informations to the FAQ?

File: using_the_compiler/README.md [hunk 2]
```diff
@@ -44,6 +44,18 @@ Make sure to always use `--release` for production-ready executables and when pe
 
 The reason for this is that performance without full optimizations is still pretty good and provides fast compile times, so you can use the `crystal` command almost as if it were an interpreter.
 
+To reduce the binary size, you can add the `--no-debug` flag and use the `strip` command. Debug symbols will be removed, use this options if only the size is an issue and you won't need to debug the program.
```
Comment (RX14) [inline using_the_compiler/README.md]: *How* is the `strip` command used?

File: using_the_compiler/README.md [hunk 3]
```diff
@@ -44,6 +44,18 @@ Make sure to always use `--release` for production-ready executables and when pe
...
+
+## Creating a standalone executable
```
Comment (RX14) [inline using_the_compiler/README.md]: Please call it "statically-linked".

File: using_the_compiler/README.md [hunk 4]
```diff
@@ -44,6 +44,18 @@ Make sure to always use `--release` for production-ready executables and when pe
...
+
+To build a standalone executable of your program:
+
+`
+$ crystal build some_program.cr --release --static
+```
+
+More informations about statically linking [can be found on the wiki](https://github.com/crystal-lang/crystal/wiki/Static-Linking)
```
Comment (RX14) [inline using_the_compiler/README.md]: I think we shouldn't include this link.

Comment (j8r) [reply using_the_compiler/README.md]: But if we don't provide further informations about known issues the users will face, they will lose their time searching answers in existing issues, or even create new ones.

Comment (RX14) [reply using_the_compiler/README.md]: Honestly the `--static` flag won't do what 99% of users think it does. They'll think static compilation -> go, and our static compilation is nothing like go's because we still link to non-crystal C libraries, whereas go doesn't. So I strongly vote to just leave the flag out until we have an acceptable static linking story on any target that *isn't* musl.

**Comment by j8r** [comment]
After discussion, the nature of this PR has quite changed and is now more about the binary size reduction (see the updated description)

**Comment by asterite** [comment]
Does `strip` exist for windows?

`strip` seems like a general utility. I'm not sure it's a good idea to talk about this in a programming language. Searching "how to reduce binary size" in Google will probably point to strip or other utilities.

**Comment by j8r** [comment]
No problem @asterite , I can just mention.

File: using_the_compiler/README.md [hunk 5]
```diff
@@ -20,7 +20,7 @@ Alternatively you can use the `run` command:
 $ crystal run some_program.cr
 ```
 
-## Creating an executable
+## Creating a dynamically-linked executable
```
Comment (mverzilli) [inline using_the_compiler/README.md]: This change is unrelated to binary size reduction, please send it in a separate PR.

File: using_the_compiler/README.md [hunk 6]
```diff
@@ -44,6 +44,12 @@ Make sure to always use `--release` for production-ready executables and when pe
 
 The reason for this is that performance without full optimizations is still pretty good and provides fast compile times, so you can use the `crystal` command almost as if it were an interpreter.
 
+## Reducing the executable's binary size
+
+This will remove debug symbols. Use this options only if the size is an issue and debugging isn't planned.
```
Comment (mverzilli) [inline using_the_compiler/README.md]: Can you send this paragraph below the one that starts with "To reduce...", it's often best to start with the core of the section ("to reduce binary size you can you the `--no-debug` flag") and then explain further ("you get a smaller binary file because you're effectively removing debug symbols").

Comment (mverzilli) [reply using_the_compiler/README.md]: "this options" => "these options"

File: using_the_compiler/README.md [hunk 7]
```diff
@@ -44,6 +44,12 @@ Make sure to always use `--release` for production-ready executables and when pe
...
+
+To reduce the binary size, the `--no-debug` flag can be added. Furthermore, commands like `strip` can also be used.
```
Comment (mverzilli) [inline using_the_compiler/README.md]: The reference to strip is a bit too loose. Is it a flag for the compiler? Is it the Unix `strip` tool? Either remove the reference or add a bit more context assuming the reader has never been exposed to `strip`.

**Comment by j8r** [comment]
I've made some changements according to your reviews @mverzilli

File: using_the_compiler/README.md [hunk 8]
```diff
@@ -44,6 +44,16 @@ Make sure to always use `--release` for production-ready executables and when pe
...
 
+## Reducing the executable's binary size
+
+To reduce the binary size by removing debug symbols, the `--no-debug` flag can be added.
+
+```                                                                                                                                      
+$ crystal build some_program.cr --no-debug
+```
+
+Use this options only if the size is an issue and debugging isn't planned.
```
Comment (mverzilli) [inline using_the_compiler/README.md]: this options => this option

Comment (j8r) [reply using_the_compiler/README.md]: Thanks, the corrections are now applied

**Comment by j8r** [comment]
Nothing more to do @mverzilli ?

**Comment by mverzilli** [comment]
Thank you!
